core/trino-parser/src/main/java/io/trino/sql/parser/AstBuilder.java

```diff
@@ -1555,30 +1558,89 @@ public Node visitDropRole(SqlBaseParser.DropRoleContext context)
                 visitIfPresent(context.catalog, Identifier.class));
     }
 
+    @Override
+    public Node visitGrantPrivileges(SqlBaseParser.GrantPrivilegesContext context)
+    {
+        List<String> privileges;
+        if (context.ALL() != null) {
+            privileges = ImmutableList.of();
+        }
+        else {
+            privileges = extractPrivileges(context.privilegeOrRole());
+        }
+        return new Grant(
+                getLocation(context),
+                privileges.isEmpty() ? Optional.empty() : Optional.of(privileges),
+                createGrantObject(getLocation(context), context.grantObject()),
+                getPrincipalSpecification(context.principal()),
+                context.OPTION() != null);
+    }
+
     @Override
     public Node visitGrantRoles(SqlBaseParser.GrantRolesContext context)
     {
+        // Check legality for a role grant.
         return new GrantRoles(
                 getLocation(context),
-                ImmutableSet.copyOf(getIdentifiers(context.roles().identifier())),
+                extractRoles(context.privilegeOrRole()),
                 ImmutableSet.copyOf(getPrincipalSpecifications(context.principal())),
                 context.OPTION() != null,
                 getGrantorSpecificationIfPresent(context.grantor()),
                 visitIfPresent(context.catalog, Identifier.class));
     }
 
+    @Override
+    public Node visitRevokePrivileges(SqlBaseParser.RevokePrivilegesContext context)
+    {
+        List<String> privileges = extractPrivileges(context.privilegeOrRole());
```

Comment:
How is this method handling "ALL PRIVILEGES" ? Is empty list a side effect of the rule matching "ALL PRIVILEGES" instead of privilegeOrRole? If so, let's make the check explicit.

Also, see my comment above re: the semantics of empty list.

Reply:
It was missing the required test for "ALL PRIVILEGES".  Added, and handled the empty list as suggested.